docs/guide/front-end-helpers.md

```diff
@@ -0,0 +1,319 @@
+| [`assign`](#assign-target-src1-src2) | Assigns properties from one or more source objects to a target object. |
+| [`attachmentUrl`](#attachmenturl-fileobj-options) | Get the file URL for an Apostrophe attachment object. |
+| [`closest`](#closest-el-selector) | Returns the closest ancestor element that matches the selector. |
+| [`emit`](#emit-el-name-data) | Emit a custom browser event on a DOM element. |
+| [`getCookie`](#getcookie-name) | Get the value of a browser cookie. |
+| [`onReadyAndRefresh`](#onreadyandrefresh-fn) | Runs the function passed in when Apostrophe refreshes page content during editing.
+| [`removeClass`](#removeclass-el-classname) | Remove a class from a DOM element, if present. |
+| [`sameSite`](#samesite-uri) | Returns `true` if the URI pass in matches the same website as the current page. |
+
+### `addClass(el, className)`
+
+Add a class to a DOM element, if missing. Often used with [`removeClass`](#removeclass-el-classname). Supports browsers without the matching native method.
+
+| Argument | What is it? |
+| -------- | ----------- |
+| `el` | a DOM element |
+| `className` | a string to be added to the `class` attribute |
+
+```javascript
+const myElement = document.querySelector('[data-my-element]');
+
+apos.util.addClass(myElement, 'is-active');
+```
+
+### `assign(target, src1, src2, ...`
+
+Assigns properties from one or more source objects to a target object. Uses [`Object.assign`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign) when available. Supports browsers without the matching native method.
+
+| Argument | What is it? |
+| -------- | ----------- |
+| `target` | an object on which to add source object properties |
+| `src1`, `src2`, etc. | one or more source objects whose properties should copy to the target object |
+
+```javascript
+const options = {
+  name: 'balloon',
+  color: 'blue'
+};
+
+apos.util.assign(options, {
+  color: 'red',
+  language: 'French'
+});
+
+console.log(options)
+// {
+//   name: 'balloon',
+//   color: 'red',
+//   language: 'French'
+// }
+```
+
+### `attachmentUrl(fileObj, options)`
+
+Get the file URL for an Apostrophe attachment object. Optionally pass an options object to get a specific version of the file. File (non-image) attachment objects include a single `_url` property already, so this is primarily used for retrieving specific versions of an image.
+
+| Argument | What is it? |
+| -------- | ----------- |
+| `fileObj` | an attachment object, such as from an attachment field value or using a template helper method to parse an [image widget](/guide/core-widgets.md#image-widget) value |
+| `options` | image options to apply for the resulting URL |
+
+| Options | What is it? |
+| -------- | ----------- |
+| `crop` | an object of image cropping coordinates with `left`, `top` and `width` properties (usually populated from the cropping UI and included on the image object) |
```

Comment:
We have no cropping UI yet. Might want to put a pin in this.

Reply:
True. I wasn't sure about that.